### PR TITLE
Add raw-audio streaming sessions for Sortformer

### LINKED ISSUE #15918: No Streaming API for streaming sortformer?
**Is your feature request related to a problem? Please describe.**

I would like an easy way to use the streaming sortformer on a stream of audio a chunk at a time.

**Describe the solution you'd like**

diarize_step() and reset() methods.

**Describe alternatives you've considered**

It is likely currently possible to use streaming_forward_step() to cobble something together, but it would be nice to have a high-level api like the diarize() method. [Edit: Unclear how to stream audio through the preprocessor].

**Additional context**

### PR DESCRIPTION
**Pull request by tianyi-zhang-02**
# What does this PR do?

Adds a stateful raw-audio streaming session for a fixed batch of independent Streaming Sortformer inputs.

**Collection**: ASR / Speaker Tasks

Fixes #15918.

# Changelog

- Keep the implementation and validation in `nemo/collections/asr/parts/utils/sortformer_utils.py`; `sortformer_diar_models.py` is unchanged from current main.
- Construct `SortformerStreamingSession(model, batch_size=N)` directly. Require an evaluation model with `streaming_mode=True`, `async_streaming=True`, and unnormalized features (`normalize="NA"`, `None`, or `False`).
- Preserve independent per-row waveform buffers, progress, finalization, and asynchronous speaker-cache state. Return padded probabilities with per-row valid output lengths.
- Reuse the checkpoint preprocessor without overriding normalization, dither, or feature padding. Evaluation already disables dither; valid feature windows exclude dense padding. Whole-recording statistical normalization is unsupported for live sessions because it requires future audio.
- Schedule only ready rows through the existing asynchronous model path so temporarily idle rows do not trigger its zero-length finalization behavior. Group equal central lengths to avoid padding dilution during output downsampling.
- Require complete non-final model chunks even with zero right context, retain sufficient STFT support for centered/exact-padding configurations, and reject audio for finalized rows before consuming any row.
- Add atomic `reset_streams(reset_mask)` lifecycle support. It accepts only finalized rows, clears their waveform/progress and all asynchronous model state, makes those slots reusable, and preserves every unselected row exactly. Whole-session `reset()` remains available.
- Update usage and add an independent full-waveform preprocessing + streaming-model equivalence oracle, with focused regression coverage for the boundaries above.

# Usage

```python
import torch
from nemo.collections.asr.parts.utils.sortformer_utils import SortformerStreamingSession

model.eval()
model.async_streaming = True
session = SortformerStreamingSession(model, batch_size=2)
probabilities, probability_lengths = session.diarize_step(
    audio_batch,
    audio_chunk_lengths=audio_lengths,
    is_final=final_mask,
)
stream_0_probabilities = probabilities[0, : probability_lengths[0]]

# After stream 0 is finalized and its final output has been consumed:
session.reset_streams(torch.tensor([True, False]))
# Stream 0 can receive a replacement recording while stream 1 remains active.
```

Input is mono float audio at the checkpoint sample rate. A finalized row may remain in subsequent calls with zero valid input length while other rows continue. `reset_streams()` validates the complete mask before changing state and rejects non-finalized rows; `reset()` restarts the complete batch. `batch_size=1` uses the same session abstraction.

# Testing

Latest tested SHA: `5dce0189f35887e9bd4afeaa896241e366b7d478`.
Synced main: `f613eed86ed4696db0891aac4e9104337a39142c`.

Environment: local macOS 26.6.2 / arm64, Python 3.12.2, PyTorch 2.12.0, pytest 9.0.3; CPU only.

- `python -m pytest -q tests/collections/speaker_tasks/test_diar_sortformer_models.py tests/collections/speaker_tasks/test_diar_sortformer_modules.py tests/collections/speaker_tasks/utils/test_sortformer_utils.py -m 'not pleasefixme' --cpu --use_local_test_data` — **348 passed** on the tested SHA.
- The row-reuse regression finalizes row 0 while row 1 remains active, snapshots row 1's waveform/progress and complete model state, resets row 0, and processes a replacement recording while row 1 continues. It exercises nonzero FIFO, speaker-cache predictions/lengths, and silence-profile state; both completed outputs match fresh independent sessions.
- The atomic-validation regression selects one finalized and one non-finalized row, requires an exception, and verifies that every waveform buffer, progress field, finalization flag, streaming-state field, and state-key set remains unchanged.
- Independent unit reference uses `model.process_signal(full_waveform, lengths)` followed by `model.forward_streaming(valid_features, feature_lengths)`; it never constructs the new session. Coverage includes low/high resolution, `pad_to=0/16`, zero/nonzero right context, irregular chunks, heterogeneous lengths, nonzero input padding, staggered finalization, compaction, extra output downsampling, exact STFT padding, idle FIFO state preservation, selective reset, whole-session reset, and input validation.
- Regression negative controls failed before their fixes: incomplete zero-right-context chunks, pooled final-row padding, idle-row FIFO mutation, partial consumption on invalid finalized input, and insufficient exact-pad STFT support. Both new lifecycle regressions fail on the parent commit because `reset_streams()` is absent.
- Official `nvidia/diar_streaming_sortformer_4spk-v2.1` checkpoint independent equivalence — **passed** on the tested SHA with `torch.testing.assert_close` default float32 tolerances. The reference again uses full-waveform `process_signal()` + valid-feature `forward_streaming()` and never constructs the session.
  - Offline cached checkpoint revision `fafaab5faa1617a0ca52d38dd3dc4bd636800d3d`; checkpoint SHA256 `8abd32832159c6ac1148c926b7276f35ba34582c444e559dce1f1253fea42ef8`.
  - CPU 4 threads / 1 interop thread; float32, seed 7, 16 kHz synthetic noisy tones (220/330 Hz), sample lengths `[51237, 33619]`.
  - 1.04 s latency configuration: chunk 6, left/right context 1/7, speaker cache 188, FIFO 188, cache update period 144; output factor 8, `normalize="NA"`, retained `pad_to=16`, and configured dither `1e-5` (inactive in eval).
  - Irregular sample chunks `[160, 997, 3200, 441, 8192, 73, 5000, 1201]`, unequal per-row sizes, temporarily and fully idle calls, 13-sample nonzero trailing input padding, and staggered finalization.
  - Complete output shapes `[(40, 4), (27, 4)]`; maximum absolute errors `[0.0, 0.0]`; pre-final emission `[24, 6]` frames. Both rows compacted their waveform buffers.
- `pre-commit run --from-ref upstream/main --to-ref HEAD` — passed.
- `git diff --check upstream/main...HEAD` — passed.
- All six commits unique to the branch have author-matching DCO sign-offs. Model-file diff against synced main is empty.

Limits: this round uses local CPU only. Current-head CUDA validation was **not run**; previous-head RTX 4090 results are historical and do not validate this SHA. Synthetic equivalence is not a DER evaluation. A full documentation build was not run for the README-only usage update.

# GitHub Actions CI

NVIDIA runner validation requires maintainer approval for this external contribution; local results above are independent of that approval.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines.
- [x] Added necessary independent equivalence and regression tests.
- [x] Updated the Streaming Sortformer usage documentation.
- [x] No optional-install component is affected.

**PR Type**:

- [x] New Feature
- [x] Bugfix
- [ ] Documentation

# Additional Information

- Related issue: #15918.
- Cancellation of non-finalized rows and dynamic batch resizing are intentionally out of scope; `reset_streams()` only reuses completed slots.
- #16210 remains a separate downstream stack. It will need to adopt direct session construction, the required async setting, and per-row reset semantics when its base is refreshed; no speaker-limit functionality is included here.

### CODE REVIEW

**Review by ipmedenn — CHANGES_REQUESTED** [review]
Thanks for working on the much-needed raw-audio streaming API.

I’m wondering whether we could build this on top of the existing asynchronous batched-streaming machinery instead of introducing a separate path specialized for `batch_size=1`.
Ideally, the session would manage a fixed batch of independent streams, with per-stream buffers, lengths, offsets, and finalization state. A single stream would then naturally be the `batch_size=1` case of the same implementation. This would also avoid the additional maintenance burden and potential behavioral drift between separate single-stream and batched paths.
If there are constraints that make the batched approach impractical here, it would be helpful to understand them. My preference would be to establish the general abstraction in this PR rather than make the single-stream API public first.

And one smaller organizational point: could the session implementation live in `utils/sortformer_utils.py` rather than introducing a separate `utils/streaming_sortformer.py` module? Since this is a Sortformer-specific inference utility, keeping it with the existing Sortformer utilities seems more consistent and avoids another narrowly scoped module.

**Comment by ipmedenn** [comment]
Hi @tianyi-zhang-02 — apologies for the delayed follow-up, and thanks for addressing the earlier feedback.
The PR now conflicts with `main`, which includes substantial Sortformer changes. Could you please sync the branch, resolve the conflicts, rerun the relevant tests, and update the test results/head SHA in the description?

I would also prefer to keep `sortformer_diar_models.py` unchanged:
- Construct `SortformerStreamingSession(model, ...)` directly instead of adding `create_streaming_session()`.
- Require and validate `model.async_streaming=True` instead of adding a per-call override to `forward_streaming_step()`.

A few preprocessing questions:
- Could we remove `normalize_batch()`? Current streaming Sortformer models use `normalize: "NA"`, so it is a no-op.
- Setting `dither=0` appears redundant after `.eval()`.
- Is `pad_to=0` required for correctness, or only to avoid extra computation?

The existing tests cover irregular chunks, staggered finalization, and buffer compaction. However, both the reference and tested paths use the new session implementation. Could you add one equivalence test comparing the completed session output against the existing full-waveform preprocessing and streaming-model path?

The fixed-batch session direction looks good; I would just keep it contained in `sortformer_utils.py` and minimize unrelated model/preprocessing changes. Thanks again for your patience.

**Comment by tianyi-zhang-02** [comment]
@ipmedenn Addressed in [e024a2c](https://github.com/NVIDIA-NeMo/Speech/commit/e024a2c658eea763fe4ca73d9a1157b051620b8e), synced with main `2224cfb6a`.

- `sortformer_diar_models.py` now has zero diff against main. Usage directly constructs `SortformerStreamingSession(model, ...)`; the utility validates `model.async_streaming=True`, with no per-call override.
- Removed the extra `normalize_batch()` call and all preprocessor overrides/copying. The session requires unnormalized features (`NA`/`None`/`False`). Source confirms eval already gates dither off; `pad_to` only adds dense padding after valid-length masking, so `pad_to=0` is unnecessary for correctness. Checkpoint settings are preserved, and equivalence covers both `pad_to=0/16`.
- The new reference never calls the session: complete-waveform `process_signal()` → valid-feature `forward_streaming()`. It covers irregular/heterogeneous chunks, staggered finalization and compaction, with additional failing-before-fix controls for zero-right-context boundaries, final-row pooling, idle FIFO state under the new main semantics, exact STFT padding, and atomic invalid-input rejection.
- On this exact SHA, all **346** relevant CPU tests pass. The official v2.1 checkpoint independently matches completed two-row output exactly (shapes `(40,4)` / `(27,4)`, maximum errors `0.0` / `0.0`). DCO, pre-commit and diff checks pass; the description records the full configuration and latest SHA. This round ran locally on CPU; current-head CUDA was not run.

**Comment by ipmedenn** [comment]
Hi @tianyi-zhang-02, thanks for addressing my concerns and updating the PR. I’d like to run a few tests on my side before approving it. I’ll get back to you early next week.

**Comment by ipmedenn** [comment]
Hi @tianyi-zhang-02 — thanks again for the thorough update and validation.

I tested the new API with a manifest-level raw-audio streaming script and confirmed that, for my tested
configuration, its completed outputs are equivalent to the existing full-recording inference path.

I did find one lifecycle issue with the fixed-batch abstraction. After a row is finalized, `diarize_step()`
correctly rejects additional audio for it, but the only available `reset()` operation resets the entire session.
Consequently, a completed row cannot be assigned a new recording while another row remains active.

For example, a rolling scheduler should be able to do:

```text
[recording 1, recording 2]
[recording 1, recording 3]  # recording 2 finished
[recording 4, recording 3]  # recording 1 finished
```

With the current API, the finished slot must instead remain idle until the other recording finishes. Calling
`reset()` earlier would discard that active row's audio progress, FIFO, speaker cache, and silence state. In offline
inference this causes head-of-line blocking for recordings of different lengths; in a long-lived service it can
progressively strand batch capacity.

Could we add an atomic per-row reset operation, such as:

```python
session.reset_streams(reset_mask)
```

I would expect it to:

- accept only finalized rows unless cancellation is explicitly supported;
- validate the complete mask before changing any state;
- clear the selected rows' audio buffers, offsets, counters, and finalized flags;
- reset all selected speaker-cache, FIFO, cache-metadata, and silence-profile state;
- preserve every unselected row exactly.

The caller would first submit and collect the old recording's final output, reset that row, and then provide the
replacement recording on the next step. The existing whole-session `reset()` can remain useful for teardown or
restarting an entire batch.

I do not think each aspect needs a separate unit test. Two focused regressions should be sufficient:

1. Finalize row 0 while row 1 remains active, reset row 0, and then process a replacement recording while continuing
   row 1. Verify that row 1 is unchanged by the reset and compare both completed recordings with fresh independent
   sessions. This covers successful reuse, active-peer preservation, and state leakage.
2. Attempt a reset mask containing both a finalized row and a non-finalized row. Require an exception and verify that
   the complete session remains unchanged, demonstrating atomic validation.

The existing tests already cover staggered finalization, idle rows, and paused-row FIFO behavior, so I would not
duplicate those scenarios further.

The current implementation otherwise behaved well in my equivalence testing. I think row-level reuse should be
addressed before exposing the fixed-batch session as the public API, since it determines whether the batch can remain
continuously occupied.

**Comment by tianyi-zhang-02** [comment]
@ipmedenn Addressed in [`5dce0189f`](https://github.com/NVIDIA-NeMo/Speech/commit/5dce0189f35887e9bd4afeaa896241e366b7d478), synced with current main `f613eed86`.

- Added `session.reset_streams(reset_mask)`. It validates the complete boolean mask before mutation, accepts only finalized rows, and runs under inference mode so inference-state tensors can be reset safely. Selected rows have their waveform buffer, offsets/counters, finalized flag, FIFO, speaker cache/predictions, cache metadata, and silence profile cleared; unselected rows are left untouched. The existing whole-batch `reset()` remains. Cancellation of non-finalized rows is intentionally out of scope.
- Added the two focused regressions you requested. The reuse test finalizes row 0 with nonzero FIFO/cache/silence state while row 1 remains active, resets row 0, processes a replacement while row 1 continues, verifies row 1's complete state exactly, and compares the replacement and continued peer with fresh independent sessions. The mixed finalized/non-finalized mask test requires an exception and verifies that the complete session remains exactly unchanged.
- Updated the usage documentation with the collect-final-output → selective-reset → replacement workflow.

On this exact SHA:

- all **348** relevant CPU tests pass;
- the official v2.1 checkpoint independent full-waveform reference still matches the session exactly (`[(40, 4), (27, 4)]`, maximum errors `[0.0, 0.0]`) under irregular chunks, idle calls, nonzero padding, staggered finalization, and buffer compaction;
- branch-wide pre-commit, diff checks, and all six DCO sign-offs pass;
- `sortformer_diar_models.py` remains identical to main.

The PR description now records the full current-head configuration and results. Current-head CUDA was not rerun; this round is CPU-only. Could you take another look when convenient? Thanks!
